trie/stacktrie.go

```diff
@@ -160,20 +156,18 @@ func (st *StackTrie) setDb(db ethdb.KeyValueWriter) {
 	}
 }
 
-func newLeaf(ko int, key, val []byte, db ethdb.KeyValueWriter) *StackTrie {
+func newLeaf(key, val []byte, db ethdb.KeyValueWriter) *StackTrie {
 	st := stackTrieFromPool(db)
 	st.nodeType = leafNode
-	st.keyOffset = ko
-	st.key = append(st.key, key[ko:]...)
+	st.key = key
```

Comment:
Here I'm not sure. Previously this worked as a copy? (the `st.key` is always empty after taking an object from the pool).

Reply:
Yes, so previously the incoming `key` was always copied, but the `append` operation made it so that we could sometimes use an existing buffer, so no alloc was needed. WIth this change, leafs will reference the same backing slice. 

We need to investigate whether that can be a problem.

Reply:
> (the st.key is always empty after taking an object from the pool).

That's because in `Reset`, which is called when returning things to the pool, we do
```
	st.key = st.key[:0]
```
That means we just truncate it, but don't dereference the backing-slice. So the actual underlying slice is reused, whenever we at a later point do `append(st.key, ...)`. 

A safer alternative to what you're doing would simply be to change it back into 
```
	st.key = append(st.key, key...)
```

Reply:
I'm running the stacktrie fuzzer on it now, let's see if it finds anything

Reply:
I now see these were for both leaf and ext node (the only ones which have keys). I can restore these. Are they only for memory management optimization unless someone would change the content underneath?

Reply:
> Are they only for memory management optimization unless someone would change the content underneath?

Not sure I understand the question. But basically, you assume that the caller will not modify the input `key` ever again. Which may be true, but I'm not sure we have any guarantee. 
The previous code always copied the `key`, but did so in a way which for the most part never caused any allocs to happen. I think the previous way to do it is better.

Reply:
Thanks you to the the of Earth Angels

Reply:
I think it's ok to direct hold the byte slice. We do have one place with modify the underlying slice, https://github.com/ethereum/go-ethereum/blob/master/trie/stacktrie.go#L437 but it's only for the leaf node.